File: Data Structures & Algorithms/jump-game/submission-1.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        """
            1. lets do brute force first think through the solution
            2. now lets cache this


        """
        cache = {}
        def helper(idx,nums: List[int]) -> bool:
            if idx >= len(nums)-1:
                return True
            
            if idx in cache:
                return cache[idx]
            
            ans = False
            for i in range(1, nums[idx]+1):
                ans |= helper(idx + i, nums)

            cache[idx] = ans
            return ans
        

        return helper(0,nums)

File: Data Structures & Algorithms/target-sum/submission-1.cpp
class Solution {
public:
    int helper(vector<int>& nums,int target,int curr,int n){
        if(curr >= n){
            return (target == 0);
        }

        int ans = 0;
        ans += helper(nums,target-nums[curr],curr+1,n);
        ans += helper(nums,target+nums[curr],curr+1,n);
        return ans;
    }

    int findTargetSumWays(vector<int>& nums, int target) {
        int n = nums.size();
        return helper(nums,target,0,n);
    }
};
